File: integerToRoman/integerToRoman.cpp
// Source : https://leetcode.com/problems/integer-to-roman/
// Author : weekend27
// Date   : 2015-12-17

/**********************************************************************************

Given an integer, convert it to a roman numeral.

Input is guaranteed to be within the range from 1 to 3999.

**********************************************************************************/

// How to do it:
// process digits according to different intervals

class Solution {
public:
    string intToRoman(int num) {
        char symbol[7] = {'I', 'V', 'X', 'L', 'C', 'D', 'M'};
        string roman;
        int scale = 1000;
        for (int i = 6; i >= 0; i -= 2){
            int digit = num / scale;
            if (digit != 0){
                if (digit <= 3){
                    roman.append(digit, symbol[i]);
                }
                else if(digit == 4){
                    roman.append(1, symbol[i]);
                    roman.append(1, symbol[i+1]);
                }
                else if(digit == 5){
                    roman.append(1, symbol[i+1]);
                }
                else if(digit <= 8){
                    roman.append(1, symbol[i+1]);
                    roman.append(digit-5, symbol[i]);
                }
                else if(digit == 9){
                    roman.append(1, symbol[i]);
                    roman.append(1, symbol[i+2]);
                }
            }
            num = num % scale;
            scale /= 10;
        }
        return roman;
    }
};

File: integerToRoman/integerToRoman.java
// Source : https://leetcode.com/problems/integer-to-roman/
// Author : weekend27
// Date   : 20156-04-26

/**********************************************************************************

Given an integer, convert it to a roman numeral.

Input is guaranteed to be within the range from 1 to 3999.

**********************************************************************************/


public class Solution {
    public String intToRoman(int num) {
        String[][] roman = {
            {"", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"},
            {"", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"},
            {"", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"},
            {"", "M", "MM", "MMM"}
        };

        StringBuilder ret = new StringBuilder();
        int digit = 0;
        while (num > 0) {
            int remain = num % 10;      // remain part
            ret = ret.insert(0, roman[digit][remain]);      // insert the remain part in the 0 index
            digit++;
            num /= 10;
        }
        String result = ret.toString();
        return result;
    }

}
